feat(git): disable sorting commits topologically (#804)

SebClapie opened the pull request:
## Description

Add sorting option for commits 

## Motivation and Context

This change is related to #804 
When I tried git-cliff on my personal git repository, the commits order was not correct. 
I finally found why the order was not correct. The CHANGELOD generated got the same order than the command git log --oneline --topo-order.


## How Has This Been Tested?

I did not add a specific test for this feature, I apologize, I just installed rust for the first time yesterday and I never used GitHub before ..
I tried to follow your recommandation.
I launched:
- cargo build
- cargo test
- cargo fmt
- cargo clippy

I also tested on my side that the sort is functional on my repro. I tested the command line option --disable-topo-order-commits
I also tested this option inside my file cliff.toml

## Screenshots / Logs (if applicable)

## Types of Changes


- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (no code change)
- [ ] Refactor (refactoring production code)
- [ ] Other <!--- (provide information) -->

## Checklist:

- [X] My code follows the code style of this project.
I launched cargo fmt and cargo clippy
- [X] I have updated the documentation accordingly.
Maybe the text can be improved
- [X] I have formatted the code with [rustfmt](https://github.com/rust-lang/rustfmt).
- [X] I checked the lints with [clippy](https://github.com/rust-lang/rust-clippy).
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

<!--- Thank you for contributing to git-cliff! ⛰️  -->

orhun reviewed (commented):
Thanks for the PR! This is definitely one of the crucial missing pieces so I appreciate you taking time to implement it.

Some comments:

- I think the option should be called `topo_order_commits`. `disable_topo_order_commits` does not read well since we are using a boolean. (e.g. disable disable_topo_order_commits)
- We already have a `topo_order` option for sorting the tags. I think we should rename that to `topo_order_tags` and deprecate the old option. What do you think? (cc. @Cyclonit since he worked on revamping the config before)
- We should add fixture tests for this, see the instructions [here](https://github.com/orhun/git-cliff/tree/main/.github/fixtures#adding-new-fixtures).

Cyclonit commented:
I would propose refactoring the system to have two options:

- `git.commit_order_by` = `topology`/`time`
- `changelog.release_order_by` = `topology`/`time`

Having enums instead of booleans improves readability significantly. But I know @orhun likes using as few words as possible ;)

However, we do not have any way of soft deprecating old config options and adding new ones beside it. That is something that would have to be designed.

SebClapie commented:
Hello,

Thanks @orhun for your feedback.
I will a look to the fixture, and I may come back to you if I need help

I can rename `disable_topo_order_commits `to `topo_order_commits`.
Or do I look for @Cyclonit proposal  ?

- `git.commit_order_by = topology/time`

orhun commented:
> I would propose refactoring the system to have two options:
> 
>     * `git.commit_order_by` = `topology`/`time`
> 
>     * `changelog.release_order_by` = `topology`/`time`

In that case we also need to deprecate the old options. I'd prefer we go with `topo_order_commits` for now and revamp the configuration later.

> However, we do not have any way of soft deprecating old config options and adding new ones beside it. That is something that would have to be designed.

We can probably add a warning while parsing the config 🤔 I'll look into it.

LMK @SebClapie when this is ready for review.

SebClapie commented:
Hello @orhun ,

I updated the code to rename the option `topo_order_commits `, and the default value is true to keep previous behavior
I added a new test fixture, but I don't know why I got this issue "line 2: ./commit.sh: Permission denied". I follow your procedure for add a new ficture, and it works locally for me.

Cyclonit commented:
> Hello @orhun ,
> 
> ... I added a new test fixture, but I don't know why I got this issue "line 2: ./commit.sh: Permission denied".

The `commit.sh` you creqted does not have POSIX permissions assogned to it. This usually happens because you created the file on windows. Run this to fix it, commit and push again:

```sh
git update-index --chmod=+x <file>
```

Cyclonit commented on .github/fixtures/test-topo-order-commits/cliff.toml:
As of #1090 you don't have to include options if they match those in https://github.com/orhun/git-cliff/blob/main/config/cliff.toml (e.g. `changelog.header` and `changelog.footer`)

SebClapie commented:
The Typos action failed. But I don't think it's linked to my modification
I don't know how to retrigger it ....

Cyclonit commented on config/cliff.toml:
```diff
@@ -94,6 +94,8 @@ link_parsers = []
 use_branch_tags = false
 # Order releases topologically instead of chronologically.
 topo_order = false
+# Topogical order for commits.
```
I'd change this comment to `Order releases topologically instead of chronologically." for it to match the one on `topo_order`.

orhun replied:
Yup!

```suggestion
# Order releases topologically instead of chronologically.
```

orhun commented on website/docs/configuration/git.md:
```diff
@@ -252,6 +252,17 @@ If set to `true`, tags are processed in topological order instead of chronologic
 
+### topo_order_commits
+
+If set to `false`, commits are processed in chronological order instead of topological.
```
We should use the same wording as `topo_order` above.

```suggestion
If set to `true`, commits are processed in topological order instead of chronological.
```

orhun reviewed (commented):
LGTM, some minor comment issues :)

Can you also create a new issue for tracking the deprecation of `topo_order` (in favor of `topo_order_tags`)?

orhun commented on .github/fixtures/test-topo-order-commits/cliff.toml:
```diff
@@ -0,0 +1,44 @@
+# Optionally sets the commit's `scope` and can decide to exclude commits from further processing.
+commit_parsers = [
+    { message = "^feat", group = "Features", default_scope = "app" },
+    { message = "^fix", group = "Bug Fixes", scope = "cli" },
+	{ message = "^test", group = "Test"},
+]
```
```suggestion
    { message = "^fix", group = "Bug Fixes", scope = "cli" },
    { message = "^test", group = "Test"},
]
```

orhun commented on website/docs/configuration/git.md:
````diff
@@ -252,6 +252,17 @@ If set to `true`, tags are processed in topological order instead of chronologic
+### topo_order_commits
+
+If set to `true`, commits are processed in topological order instead of chronological.
+
+
+```toml
+# if false , sorting commit is equivalent to git log 
````
```suggestion

```toml
# if false, sorting commit is equivalent to git log 
```

orhun reviewed (commented):
Some styling suggestions

Cyclonit commented:
> > Some minor things left :)
> > 
> > * The fixture config can be simplified.
> > * A small tweak in docs is needed (see my comment)
> > * We need an issue to track the deprecation of config
> 
> Hello @orhun what do you mean about the deprecation of config ? You want to replace `topo_order` by `topo_order_tags `?

The current config schema has grown historically and is in need of a refactor. I attempted to do this back in #541, but couldn't keep up with other changes at the time. The goal of improving the config in general remains though.

One of the requirements for overhauling the config is the ability to soft deprecate old options. Otherwise any change to the config would be an immediate breaking change.

I don't think you need to care about that right now.

orhun reviewed (approved):
LGTM, thanks!

We'll probably talk about the deprecation of some options during the config revamp, which should be git-cliff v3 eventually.
